ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/JumpingGovernor.hs

```diff
@@ -0,0 +1,337 @@
+          Jumper nextJumpVar _ Happy -> writeTVar nextJumpVar $ Just $! castPoint (headPoint dynamoFragment)
+          _ -> pure ()
+
+processJumpResult ::
+  ( MonadSTM m,
+    Ord peer, HasHeader blk, HasHeader (Header blk)
+  ) =>
+  Handle m peer blk ->
+  peer ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult governor@Handle {clientStates, clientHandles} peer jumpResult =
+  ((Map.! peer) <$> readTVar clientStates) >>= \case
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment governor
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
+            chainSyncStateVar <- cschState . (Map.! peer) <$> readTVar clientHandles
+            modifyTVar chainSyncStateVar $ \chainSyncState -> chainSyncState {csCandidate = slicedDynamoFragment}
+            modifyTVar clientStates . Map.insert peer $ Jumper nextJumpVar goodPoint' Happy
+          (Happy, RejectedJump badPoint) ->
+            -- The previously happy peer just rejected a jump. We look for an
+            -- intersection.
+            lookForIntersection nextJumpVar goodPoint badPoint
+          (LookingForIntersection badPoint, AcceptedJump goodPoint') ->
+            -- The peer that was looking for an intersection accepts the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint' badPoint
+          (LookingForIntersection _, RejectedJump badPoint') ->
+            -- The peer that was looking for an intersection rejects the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint badPoint'
+  where
+    lookForIntersection nextJumpVar goodPoint badPoint = do
+      (dynamoFragment0 :: AnchoredFragment (Header blk)) <- fromJust <$> getDynamoFragment governor -- full fragment
+      let (dynamoFragment1 :: AnchoredFragment (Header blk)) = snd $ fromJust $ splitAfterPoint dynamoFragment0 goodPoint -- fragment from goodPoint (excl)
+      let (dynamoFragment2 :: AnchoredFragment (Header blk)) = fst $ fromJust $ splitBeforePoint dynamoFragment1 badPoint -- fragment from goodPoint to badPoint (excl)
+      case (toOldestFirst dynamoFragment2 :: [Header blk]) of
+        [] -> do
+          modifyTVar clientStates . Map.insert peer $ Jumper nextJumpVar goodPoint FoundIntersection
+          electNewObjector governor
+        points -> do
+          let (middlePoint :: Header blk) = points !! (length points `div` 2)
```

Comment:
Can't this be accomplished better with the `AnchoredFragment` representation?

Reply:
Possibly! And I'd find that more satisfactory tbh; I was mostly trying to get things done to have a draft. I will look into cleaning this up.